api/currency_conversion_test.go

```diff
@@ -0,0 +1,341 @@
+	"net/url"
+	"strconv"
+	"time"
+
+	"github.com/alphagov/paas-billing/db"
+	"github.com/alphagov/paas-billing/db/dbhelper"
+	"github.com/alphagov/paas-billing/fixtures"
+	"github.com/alphagov/paas-billing/server"
+	"github.com/labstack/echo"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+var _ = Describe("Currency Conversion", func() {
+	var (
+		connstr       string
+		sqlClient     *db.PostgresClient
+		authenticator = AuthenticatedNonAdmin
+		start         = time.Date(2018, 3, 1, 0, 0, 0, 0, time.UTC)
+	)
+
+	ExpectJSON := func(actual interface{}, expected interface{}) {
+		b1, err := json.Marshal(actual)
+		Expect(err).ToNot(HaveOccurred(), "ExpectJSON failed for expected value")
+		b2, err := json.Marshal(expected)
+		Expect(err).ToNot(HaveOccurred(), "ExpectJSON failed for actual value")
+		Expect(string(b1)).To(MatchJSON(string(b2)))
+	}
+
+	doRequest := func(path string, v interface{}, params map[string]string) {
+		u, err := url.Parse(path)
+		Expect(err).ToNot(HaveOccurred())
+
+		q := u.Query()
+		for k, v := range params {
+			q.Set(k, v)
+		}
+		u.RawQuery = q.Encode()
+		req, err := http.NewRequest("GET", u.String(), nil)
+		Expect(err).ToNot(HaveOccurred())
+		req.Header.Set(echo.HeaderAccept, echo.MIMEApplicationJSONCharsetUTF8)
+		req.Header.Set(echo.HeaderAuthorization, FakeBearerToken)
+
+		e := server.New(sqlClient, authenticator, nil)
+		rec := httptest.NewRecorder()
+		e.ServeHTTP(rec, req)
+
+		res := rec.Result()
+		body, _ := ioutil.ReadAll(res.Body)
+		Expect(res.StatusCode).To(Equal(http.StatusOK), string(body))
+
+		err = json.Unmarshal(body, v)
+		Expect(err).ToNot(HaveOccurred(), fmt.Sprintf("failed to unmarshal json: %s\nbody: %s", err, string(body)))
+	}
+
+	addCurrency := func(code string, rate float64, validFrom time.Time) {
+		currencyRateFixtures := fixtures.CurrencyRates{{
+			Code:      code,
+			Rate:      rate,
+			ValidFrom: validFrom,
+		}}
+		err := currencyRateFixtures.Insert(sqlClient)
+		Expect(err).ToNot(HaveOccurred())
+	}
+
+	request := func(path string, from time.Time, to time.Time, v interface{}) {
+		doRequest(path, v, map[string]string{
+			"from": from.Format(time.RFC3339Nano),
+			"to":   to.Format(time.RFC3339Nano),
+		})
+	}
+
+	emptyOrgsFixture := func(guid string, spaceGUID string) fixtures.Orgs {
+		gbpOrgsFixtures := fixtures.Orgs{}
+		gbpOrgsFixtures[guid] = fixtures.Org{}
+		gbpOrgsFixtures[guid][spaceGUID] = fixtures.Space{}
+		return gbpOrgsFixtures
+	}
+
+	addEvent := func(fixture fixtures.Space, eventTime time.Time) fixtures.Space {
+		newAppEvents := append(fixture.AppEvents, fixtures.AppEvent{
+			AppGuid:               "00000001-0002-0001-0000-000000000000",
+			State:                 "STARTED",
+			InstanceCount:         1,
+			MemoryInMBPerInstance: 100,
+			Time: eventTime,
+		})
+		newServiceEvents := append(fixture.ServiceEvents, fixtures.ServiceEvent{
+			ServiceInstanceGuid: "00000001-0002-0002-0000-000000000000",
+			State:               "SPAM",
```

Comment:
`SPAM` was actually added as a temporary demo name during pairing. We found that we didn't get any rows back without a service event. This might be more intention-revealing if it were renamed.

Reply:
Wow, can this be a huge bug? Is it possible we won't get reports for organisations who never used a service?

Reply:
I would update the commit to explain why the additional event is there. 

It is a bug in the existing code, not related to this PR.

Reply:
Added a explicit commit and a FIXME for this.